<a href="https://colab.research.google.com/github/NadiaHolmlund/BDS_M1_Exam/blob/main/BDS_M1_Exam_Notes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import libraries for examples

In [ ]:
import pandas as pd
from sklearn.datasets import make_classification

# Import datasets for examples

In [ ]:
df_courses = pd.read_csv('https://raw.githubusercontent.com/HamidBekamiri/Data_Science_Handbook/main/udemy_courses_info.csv')
df_courses_year= pd.read_csv('https://raw.githubusercontent.com/HamidBekamiri/Data_Science_Handbook/main/udemy_courses_year.csv')
df_courses_2017 = pd.read_csv('https://raw.githubusercontent.com/HamidBekamiri/Data_Science_Handbook/main/df_courses_2017.csv')

In [ ]:
df_courses.head(1)

course_id                                       course_title  \
0    1006314  Financial Modeling for Business Analysts and C...   

                                                 url  is_paid  price  \
0  https://www.udemy.com/financial-modeling-for-b...     True     45   

   num_subscribers  num_reviews  num_lectures               level  \
0             2174         74.0          51.0  Intermediate Level   

   content_duration           subject  
0               2.5  Business Finance

In [ ]:
df_courses_year.head(1)

course_id  year
0    1006314  2016

In [ ]:
df_courses_2017.head(1)

course_id                        course_title  \
0    1070968  Ultimate Investment Banking Course   

                                                 url  is_paid  price  \
0  https://www.udemy.com/ultimate-investment-bank...     True    200   

   num_subscribers  num_reviews  num_lectures       level  content_duration  \
0             2147         23.0          51.0  All Levels               1.5   

            subject  year  revenue  
0  Business Finance  2017   429400

In [ ]:
print(df_courses.shape)
print(df_courses_year.shape)
print(df_courses_2017.shape)

(2959, 11)
(2959, 2)
(713, 13)


# 1. What is the best way to print the number of values in a column?

In [ ]:
df_courses['course_id'].value_counts().sum()

2959

In [ ]:
# Alternative but not as neat
df_courses['course_id'].shape

(2959,)

# 2.	What is the best way to print the highest and lowest value of a column?

In [ ]:
print(df_courses['price'].max())
print(df_courses['price'].min())

200
0


In [ ]:
# Alternative but not as neat, includes also count, mean, percentiles etc.
print(df_courses['price'].describe())

count    2959.000000
mean       63.773234
std        59.121061
min         0.000000
25%        20.000000
50%        45.000000
75%        95.000000
max       200.000000
Name: price, dtype: float64


# 3.	How do you combine two data sets?

- .merge(): Combining data on common columns or indices.
- .concat(): Combining DataFrames across rows or columns.
- .join(): Combining data on a key column or an index.

## .merge()
The datasets are merged based on column 'course_id' or indices, i.e. the number of rows remain while the number of columns increase

(the column 'year' is added, since this is the only column that differs between the datasets)

Note: () brackets

In [ ]:
df_courses_merge = pd.merge(df_courses, df_courses_year)
df_courses_merge.shape

(2959, 12)

## .concat()

The datasets are concatenated based on columns and rows, i.e. the number of rows increase and the number of columns increase

(the column 'revenue' is added, since this is the only column that differs between the datasets. The column has value NaN in the rows from df_courses_merge while the rows from df_courses_2017 is added below)

Note: ([]) brackets

In [ ]:
df_courses_concat = pd.concat([df_courses_merge, df_courses_2017])
df_courses_concat.shape

(3672, 13)

## .join()
- Inner Join: returns a dataframe with only those rows that have common characteristics.
- Full Join: returns all those records which either have a match in the left or right dataframe (missing values will be NaN on both left and right). Also called outer join.
- Left Join: returns a dataframe containing all the rows of the left dataframe ((missing values will be NaN on the right)
- Right Join: returns a dataframe containing all the rows of the right dataframe (missing values will be NaN on the left)

Note: () brackets

Note: Cannot run example on current datasets

Syntax: DataFrame.join(other, on=None, how='left', lsuffix='', rsuffix='', sort=False)

# 4.	What is the purpose of groupby operations?

Groupby operations are used to group large amounts of data and compute operations on these groups. Groupby is a 3-step process:

Splitting : It is a process in which we split data into groups by applying some conditions on datasets.

Applying : It is a process in which we apply a function to each group independently (e.g. count, sum, agg)

Combining : It is a process in which we combine different datasets after applying groupby and results into a data structure

Note: ([]] brackets + [] brackets

In [ ]:
# What is the number of courses within each subject?
df_courses_concat.groupby(['subject'])['course_id'].count()

subject
Business Finance       1191
Graphic Design          602
Musical Instruments     680
Web Development        1199
Name: course_id, dtype: int64

In [ ]:
# What is the number of courses in each level of course within each subject?
df_courses_concat.groupby(['subject','level'])['course_id'].count()

subject              level             
Business Finance     All Levels            693
                     Beginner Level        339
                     Expert Level           31
                     Intermediate Level    128
Graphic Design       All Levels            298
                     Beginner Level        242
                     Expert Level            5
                     Intermediate Level     57
Musical Instruments  All Levels            276
                     Beginner Level        296
                     Expert Level            7
                     Intermediate Level    101
Web Development      All Levels            658
                     Beginner Level        391
                     Expert Level           15
                     Intermediate Level    135
Name: course_id, dtype: int64

In [ ]:
# What is the number of subscribers in each level of course within each subject?
df_courses_concat.groupby(['subject','level'])['num_subscribers'].sum()

subject              level             
Business Finance     All Levels            1047208
                     Beginner Level         647007
                     Expert Level            30146
                     Intermediate Level     144350
Graphic Design       All Levels             688332
                     Beginner Level         315757
                     Expert Level             1008
                     Intermediate Level      58051
Musical Instruments  All Levels             328170
                     Beginner Level         429030
                     Expert Level             5431
                     Intermediate Level      84058
Web Development      All Levels            4808081
                     Beginner Level        2660049
                     Expert Level            13611
                     Intermediate Level     455546
Name: num_subscribers, dtype: int64

In [ ]:
# What is the number of courses which have been provided each year?
df_courses_concat.groupby(['year'])['course_id'].count()

year
2011       5
2012      45
2013     201
2014     490
2015    1014
2016    1204
2017     713
Name: course_id, dtype: int64

In [ ]:
# What is the number of free courses in each subject?
# Every subject category has free or paid courses, that are segregated by course id.
df_courses_concat.groupby(['subject','is_paid'])['course_id'].count()

subject              is_paid
Business Finance     False        96
                     True       1095
Graphic Design       False        35
                     True        567
Musical Instruments  False        46
                     True        634
Web Development      False       133
                     True       1066
Name: course_id, dtype: int64

In [ ]:
# What are the number of courses and the number of subscribers in each level of course within each subject?
df_courses_concat.groupby(['subject','level'])['course_id','num_subscribers'].agg({'course_id':'count', 'num_subscribers':'sum'})

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


course_id  num_subscribers
subject             level                                         
Business Finance    All Levels                693          1047208
                    Beginner Level            339           647007
                    Expert Level               31            30146
                    Intermediate Level        128           144350
Graphic Design      All Levels                298           688332
                    Beginner Level            242           315757
                    Expert Level                5             1008
                    Intermediate Level         57            58051
Musical Instruments All Levels                276           328170
                    Beginner Level            296           429030
                    Expert Level                7             5431
                    Intermediate Level        101            84058
Web Development     All Levels                658          4808081
                    Beginner Level            391          2660049
                    Expert Level               15            13611
                    Intermediate Level        135           455546

# 5.	What issues are typically associated with imbalanced data, and how do you handle it?

Imbalanced data means that the frequency of data points of one class is much higher (majority class) than of the other class (minority class).

Building a classifier on imbalanced data means it will perform well on the majority class and perform poorly on the minority class.

Typical issue in:
- Disease diagnosis
- Customer churn prediction
- Fraud detection
- Natural disaster

Example: Cancer screening where 95/100 observations do not have cancer and 5/100 do. The model may classify that all 100 do not have cancer due to the overweight of this class. Accuracy is 95% but the error is critical.

Methods to handle imbalanced data:
- Resampling (over- and undersampling)
- Synthetic Minority Oversampling Technique (SMOTE)
- Balanced Bagging Classifier
- Threshold Moving

## Resampling (over- and undersampling)

Oversampling: oversampling the minority class using replacement.

Undersampling: undersampling the majority class by removing random rows

Once both classes have a similar size, it is assumed the classifier will give equal importance to both classes.

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/0*FeIp1t4uEcW5LmSM.png)

## Synthetic Minority Oversampling Technique (SMOTE)

In SMOTE new instances are synthesized from the existing data. SMOTE looks into minority class instances and uses k nearest neighbor to select a random nearest neighbor, and a synthetic instance is created randomly in the feature space.

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*XvKkuORUzmJaz042NgB0oA.png)

## Balanced Bagging Classifier

This classifier takes two special parameters:

- Sampling_strategy: decides the type of resampling required, e.g.:
-- majority: resample only the majority class
-- all: resample all classes

- Replacement: decides whether it is going to be a sample with replacement or not.

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/The-Balanced-Bagging-flow.ppm.png)

## Threshold Moving

By default, class membership is often based on a 0.5 threshold (i.e. >0.5 belongs to one class and <0.5 belong to the other).

For imbalanced data, the threshold may be changed to an optimum value to efficiently separate two classes. To find the threshold one can use:
- ROC curve
- Precision-Recall curves
- Grid Search

# 6.	When should you use classification and when regression? SML

Use classification for discrete values and regression for continuous values

![](https://raw.githubusercontent.com/NadiaHolmlund/M1_exam/main/Classification-vs-Regression.jpg)

## Classification algorithms


Used to predict/classify discrete values such as Male or Female, True or False, Spam or Not Spam, etc.

The task of the classification algorithm is to find the mapping function to map the input(x) to the discrete output(y).

Types of ML Classification Algorithms:
- Logistic Regression (Elastic net)
- K-Nearest Neighbours
- XGBclassifier
- Support Vector Machines
- Kernel SVM
- Naïve Bayes
- Decision Tree Classification
- Random Forest Classification

## Regression algorithms


Used to predict the continuous values such as price, salary, age, etc.

The task of the regression algorithm is to find the mapping function to map the input variable(x) to the continuous output variable(y).

Types of Regression Algorithms:
- XGBregressor
- Simple Linear Regression (OLS)
- Multiple Linear Regression
- Polynomial Regression
- Support Vector Regression
- Decision Tree Regression
- Random Forest Regression

# 7.	What is the difference between Supervised Learning and Unsupervised Learning?

## Supervised ML: labelled data


- Classification: use an algorithm to accurately assign test data into specific categories (discrete values)
- Regression: uses an algorithm to understand the relationship between dependent and independent variables (continuous values)

## Unsupervised ML: unlabelled data


- Clustering: grouping unlabelled data based on their similarities or differences
- Association: uses different rules to find relationships between variables in a given dataset (e.g. recommender systems)
- Dimensionality reduction: used when the number of features (or dimensions) in a given dataset is too high. Reduces the number of data inputs to a manageable size while also preserving the data integrity.

## Goals of ML:


- Supervised: predict outcomes for new data
- Unsupervised: get insights from large volumes of new data

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-07%20at%2021.48.06.png)

# 8.	How do you evaluate performance when training a regression model?

- R Square / Adjusted R Square
- Mean Square Error (MSE) / Root Mean Squared Error (RMSE)
- Mean Absolute Error(MAE)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/R-squared-formula-linear-regression-model-640x271.jpg)
![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*5OQunI-NR-S0gAZFIit1Rw.png)

## R Square / Adjusted R Square

R sqaure measures how much variability in the dependant variable that can be explained by the model.

R Square measures how much variability in dependent variable can be explained by the model. It is the square of the Correlation Coefficient(R) and that is why it is called R Square.

R Square is calculated by the sum of squared of prediction error divided by the total sum of the square which replaces the calculated prediction with mean. R Square value is between 0 to 1 and a bigger value indicates a better fit between prediction and actual value.

Adjusted R Square will penalize additional independent variables added to the model and adjust the metric to prevent overfitting issues. 

From the sample model, we can interpret that around 79% of dependent variability can be explained by the model, and adjusted R Square is roughly the same as R Square meaning the model is quite robust.

## Mean Square Error (MSE) / Root Mean Squared Error (RMSE)

While R Square is a relative measure of how well the model fits dependent variables, Mean Square Error is an absolute measure of the goodness for the fit.
MSE is calculated by the sum of square of prediction error which is real output minus predicted output and then divide by the number of data points. It gives you an absolute number on how much your predicted results deviate from the actual number. You cannot interpret many insights from one single result but it gives you a real number to compare against other model results and help you select the best regression model.
Root Mean Square Error(RMSE) is the square root of MSE. It is used more commonly than MSE because firstly sometimes MSE value can be too big to compare easily. Secondly, MSE is calculated by the square of error, and thus square root brings it back to the same level of prediction error and makes it easier for interpretation.

## Mean Absolute Error(MAE)

Mean Absolute Error(MAE) is similar to Mean Square Error(MSE). However, instead of the sum of square of error in MSE, MAE is taking the sum of the absolute value of error.
Compare to MSE or RMSE, MAE is a more direct representation of sum of error terms. MSE gives larger penalization to big prediction error by square it while MAE treats all errors the same.

# 8(2) How do you evaluate performance when training a classification model?

- Precision - Positive predictive value (PPV)
- Recall - True positive rate (TPR)
- Accuracy - ACC
- Specificity - True negative rate (TNR)
- F1 Score - Weighted avg. of TPR and PPV
- Confusion Matrix - 2x2 with TP, TN, FP, FN
- AUC ROC curve - (Area Under The Curve) + (Receiver Operating Characteristics)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/PPV.png) Precision - Positive predictive value (PPV) 

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/TPR.png) Recall - True positive rate (TPR) 

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/ACC.png) Accuracy - ACC 

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/TNR.png) Specificity - True negative rate (TNR)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/F1.png) F1 Score - Weighted avg. of TPR and PPV 

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/confusion%20matrix.png)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/ROC.png)

# How do you evaluate performance in lecture notebooks

score (determine which score type to use, e.g. accuracy, MSE etc.)

Confusion matrix

Classification report
- Precision
- Recall
- F1 score

# 9.	How can you treat missing values?

## Why Do We Need To Care About Handling Missing Data?

- Most ML algorithms fail on missing values (except K-nearest and Naive Bayes)
- Incorrect handling of missing values may lead to bias, i.e. incorrect results
- Missing values can lead to lack of precision

## Types of missing values

- Missing Completely At Random (MCAR): Missing values are completely independent of other data. No pattern.
- Missing At Random (MAR): The probability of data missing depends only on the observed data. Pattern.
- Missing Not At Random (MNAR): There is some pattern in missing data but other observed data can not explain it. Pattern but it's unknown.

## How To Handle Missing Values?

### Deleting missing values
- Deleting the entire row
- Deleting the entire column

Pros: a model trained with the removal of all missing values creates a robust model

Cons: Loss of informations Works poortly is the % of missing values is excessive.

### Imputing missing values
- Replacing With Arbitrary Value - make an educated guess about the missing value and replace it with an arbitrary (random) value 
- Replacing With Mean - average
- Replacing With Mode - most frequent occuring value
- Replacing With Median - middle value in a sortet dataset
- Replacing with Previous Value – Forward Fill
- Replacing with Next Value – Backward Fill
- Interpolation - generating points between given points (‘polynomial’, ‘linear’, ‘quadratic’)

Pros: prevent data loss which results in deletion of rows or columns & works well w/ small data set and easy to implement

Cons: Works only with numerical continous variables. Can casue data leakage & does not factor the covariance between features

### Imputing missing values for categorical features
- Impute the Most Frequent Value
- Impute the Value “missing”, which treats it as a Separate Category

Pros: prevent data loss which results in deletion of rows or columns. Works well w/small dataset and easy to implement. Negates the loss of data by adding a unique category

Cons: Works only w/categorical variables Addition of new features to the model while encoding may result in poor performance

## Imputation using sci-kit learn library

Univariate Approach:
- only a single feature is taken into consideration. You can use the class SimpleImputer and replace the missing values with mean, mode, median or some constant value.

Multivariate Approach:
- more than one feature is taken into consideration. There are two ways to impute missing values considering the multivariate approach. Using KNNImputer or IterativeImputer classes

Nearest Neighbors Imputations (KNNImputer):
- Missing values are imputed using the k-Nearest Neighbors approach where a Euclidean distance is used to find the nearest neighbors.

Multivariate imputations (IterativeImputer):
- A strategy for imputing missing values by modeling each feature with missing values as a function of other features in a round-robin fashion.

Adding missing indicator to encode “missingness” as a feature:
- In some cases, while imputing missing values, you can preserve information about which values were missing and use that as a feature.
Because sometimes there may be a relationship between the reason for missing values (also called the “missingness”) and the target variable you are trying to predict.

# 10.	What do you understand by Precision and Recall?

Precision:
- Expresses the proportion of data points the model says were relevant that actually were relevant.
- How many retrieved elements are relevant?

Recall:
-  Expresses the proportion of actual positives the model classified correctly.
- How many relevant items are retrieved?

Often inverse relationship with each other.

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-08%20at%2000.27.54.png)

# 11.	Can you cite some examples where false positives are more problematic than false negatives?

Examples:
1. A prenatal test comes back positive for Down’s Syndrome or a disease, when the fetus does not have the disorder. Might cause some people to terminate the pregnancy
2. False positive Covid tests, which forced people to isolate for 10 days
2. When credit cards gets blocked for theft or unusual behaviour, when it is in fact not. May prevent people from payments due


# 12.	What are Recommender Systems?

A recommender system seeks to predict the rating or the preference a user might give to an item, i.e. it is an algorithm that suggests relevant items to user. (products, content, advertisements, information etc).

Methods:
- Collaborative filtering
- Content based filtering

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Content-based-filtering-vs-Collaborative-filtering-Source.png)

# 13.	What is Collaborative filtering?

Collaborative filtering:
- Based solely on  past interactions recorded between users and items in order to produce new recommendations.
-- Explicit interaction: e.g. rating on a movie
-- Implicit interaction: e.g. watching a movie
- Uses similarities between users and items simultaneously to provide recommendations
- Interactions are stored in the so-called “user-item interactions matrix”.

Classical collaborative filtering approaches:
- Memory based methods
-- user-user
-- item-item
- Model based approach
-- matrix factorisation

Cold start problem:
- Inefficient to recommend to new users
- Inefficient to recommend a new item to any users
- Inefficient to recommend items with few interactions

Solutions:
- Random strategy: recommending random items to new users or new items to random users
- Maximum expectation strategy: recommending popular items to new users or new items to most active users
- exploratory strategy: recommending a set of various items to new users or a new item to a set of various users
- Using a non collaborative method for the early life of the user or the item.

# 13(2) What is Content based filtering?

Content based filtering:
- rely on user-item interactions + additional information about users and/or items. e.g. gender, job, genres, authors, actors, duration etc.
- Build a model based on available 'features', that explain the observed user-item interactions.

Classical collaborative filtering approaches:
- Various classification models
- Various regression models

Cold start problem:
- Suffers less than collaborative, new users or items can be described by their characteristics (content) and so relevant suggestions can be done for these new entities
- Only new users or items with previously unseen features will suffer, but with older/bigger systems this is unlikely to happen.

# 14.	How can outlier values be treated?

Few outliers can ruin an ML algorithms performance, hence the importance of detection and correct handling.

## Detecting outliers

Methods:
- Boxplot
- Histogram
- Mean and Standard Deviation
- IQR (Inter Quantile Range):
-- Criteria: data points that lie 1.5 times of IQR above Q3 and below Q1 are outliers.
- Z-score Percentile:
-- Criteria: any data point whose Z-score ((Xi-mean)/std) falls out of 3rd standard deviation is an outlier.


## Treating outliers

- Trimming / removing the outlier: remove the outliers from the dataset
- Quantile based flooring and capping: the outlier is capped at a certain value above the 90th percentile value or floored at a factor below the 10th percentile value. The data points that are lesser than the 10th percentile are replaced with the 10th percentile value and the data points that are greater than the 90th percentile are replaced with 90th percentile value
- Mean/Median imputation: As the mean value is highly influenced by the outliers, it is advised to replace the outliers with the median value.
- Binning the data and categorizing them will totally avoid the outliers. It will make the data categorical instead.
- Consider them as null values and fill them up using techniques of filling up null values

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/IQR-1.png)
![](https://raw.github.com/NadiaHolmlund/M1_exam/main/440971.png)

# 15.	What do we use dimensionality reduction for?

In ML, dimensionality simply refers to the number of features/variables in the dataset. Dimensionality reduction is the task of reducing the number of features/variables without losing information and explainability.

Curse of dimenstionality:
When the number of features is high compared to number of observations
- Makes it hard for certain algorithms to train effective models


## Why is DR important

- HD often lead to over-fitting (perform well on training, poorly on test)
- HD can hamper the accuracy of algorithms because it includes irrelevant features (HD features are often correlated, hence redundant)
- DR reduces the number features, while preserving most of the relevant information
- DR makes it easier to visualize, especially if reduced to 2D or 3D
- DR avoids the curse of dimensionality
- DR reduces the time and storage space required



## Most used DR techniques:

- PCA
- Truncated SVD
- NMF
- UMAP
- t-SNE
- (M)CA
- LDA

Principal Component Analysis (PCA):
- Used to summarize information contained in continuous (i.e, quantitative) multivariate data by reducing the dimensionality of the data without losing important information.

Truncated SVD
- PCA is (truncated) SVD on centered data (by per-feature mean substraction). If the data is already centered, those two classes will do the same.
- In practice TruncatedSVD is useful on large sparse datasets which cannot be centered without making the memory usage explode.

Non-negative matrix factorization (NMF or NNMF):
- Similar to PCA but only to be used on positive values and Unlike PCA, the representation of a vector is obtained in an additive fashion. Identifies human-interpretable “patterns” or “topics”.

Uniform Manifold Approximation and Projection (UMAP):
- Recent scalable approach for visualization and general non-linear dimension reduction.

t-SNE (replaced by UMAP)

(Multiple) Correspondence Analysis ((M)CA):
- An extension of the principal component analysis suited to analyse a large contingency table formed by two qualitative variables (or categorical data).

Latent Dirichlet Allocation (LDA):
- Generative probabilistic model for collections of discrete dataset such as text corpora, where abstract topics need to be found.

# 16.	Explain clustering and some examples of typical algorithms.

Clustering groups objects in a way where objects in the same cluster are more similar to each other than to objects in other clusters.

Subgroups of clustering:
- Hard clustering: 1 observation belongs to 1 cluster (possibility of not falling into a cluster at all)
- Soft clustering: 1 observations can belong to several clusters with some probability.

## Cluster models (how the algorithm form clusters)

Connectivity-based clustering:
- Data points that are closer in the data space are more related (similar) than to data points farther away.

Centroid-based clustering:
- Clusters are represented by a central vector or a centroid. This centroid might not necessarily be a member of the dataset. (e.g. k-means)

Distribution-based clustering:
- Groups data based on their likely hood of belonging to the same probability distribution (Gaussian, Binomial, etc.)

Density-based methods:
- Groups data based on the idea that a cluster in a data space is a contiguous region of high point density, separated from other such clusters by contiguous regions of low point density.

## Clustering algorithms

K-Means:
- K-Means Clustering may be the most widely known clustering algorithm and involves assigning examples to clusters in an effort to minimize the variance within each cluster. Uses the Euclidean distances between the points as a criterion for cluster formation

Mini-Batch K-Means:
- Mini-Batch K-Means is a modified version of k-means that makes updates to the cluster centroids using mini-batches of samples rather than the entire dataset, which can make it faster for large datasets, and perhaps more robust to statistical noise.

Affinity Propagation:
- Affinity Propagation involves finding a set of exemplars that best summarize the data.

Agglomerative Clustering:
- Agglomerative clustering involves merging examples until the desired number of clusters is achieved.

BIRCH:
- BIRCH Clustering (BIRCH is short for Balanced Iterative Reducing and Clustering using
Hierarchies) involves constructing a tree structure from which cluster centroids are extracted.

DBSCAN:
- DBSCAN Clustering (where DBSCAN is short for Density-Based Spatial Clustering of Applications with Noise) involves finding high-density areas in the domain and expanding those areas of the feature space around them as clusters.

Mean Shift:
- Mean shift clustering involves finding and adapting centroids based on the density of examples in the feature space.

OPTICS:
- OPTICS clustering (where OPTICS is short for Ordering Points To Identify the Clustering Structure) is a modified version of DBSCAN described above.

Spectral Clustering:
- Spectral Clustering is a general class of clustering methods, drawn from linear algebra.

Mixture of Gaussians:
- A Gaussian mixture model summarizes a multivariate probability density function with a mixture of Gaussian probability distributions as its name suggests.

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*oNt9G9UpVhtyFLDBwEMf8Q.png)

# 17.	How do you select important variables while working on a data set?

Select variables by feature selection or feature extraction

## Filter methods

Filter methods pick up the intrinsic properties of the features measured via univariate statistics instead of cross-validation performance. These methods are faster and less computationally expensive than wrapper methods.

Information Gain:
- Information gain calculates the reduction in entropy from the transformation of a dataset. It can be used for feature selection by evaluating the Information gain of each variable in the context of the target variable.

Chi-square Test:
- The Chi-square test is used for categorical features in a dataset. We calculate Chi-square between each feature and the target and select the desired number of features with the best Chi-square scores. In order to correctly apply the chi-squared in order to test the relation between various features in the dataset and the target variable, the following conditions have to be met: the variables have to be categorical, sampled independently and values should have an expected frequency greater than 5.

Fisher’s Score:
- Fisher score is one of the most widely used supervised feature selection methods. The algorithm which we will use returns the ranks of the variables based on the fisher’s score in descending order. We can then select the variables as per the case.

Correlation Coefficient:
- Correlation is a measure of the linear relationship of 2 or more variables. Through correlation, we can predict one variable from the other. The logic behind using correlation for feature selection is that the good variables are highly correlated with the target. Furthermore, variables should be correlated with the target but should be uncorrelated among themselves.
- If two variables are correlated, we can predict one from the other. Therefore, if two features are correlated, the model only really needs one of them, as the second one does not add additional information. We will use the Pearson Correlation here.

Variance Threshold:
- The variance threshold is a simple baseline approach to feature selection. It removes all features which variance doesn’t meet some threshold. By default, it removes all zero-variance features, i.e., features that have the same value in all samples. We assume that features with a higher variance may contain more useful information, but note that we are not taking the relationship between feature variables or feature and target variables into account, which is one of the drawbacks of filter methods.

Mean Absolute Difference (MAD):
- ‘The mean absolute difference (MAD) computes the absolute difference from the mean value. The main difference between the variance and MAD measures is the absence of the square in the latter. The MAD, like the variance, is also a scale variant.’ This means that higher the MAD, higher the discriminatory power.

Dispersion ratio:
- ‘Another measure of dispersion applies the arithmetic mean (AM) and the geometric mean (GM). For a given (positive) feature Xi on n patterns. Higher dispersion implies a higher value of Ri, thus a more relevant feature. Conversely, when all the feature samples have (roughly) the same value, Ri is close to 1, indicating a low relevance feature.’

## Wrapper methods

Wrappers require some method to search the space of all possible subsets of features, assessing their quality by learning and evaluating a classifier with that feature subset. The feature selection process is based on a specific machine learning algorithm that we are trying to fit on a given dataset. It follows a greedy search approach by evaluating all the possible combinations of features against the evaluation criterion. The wrapper methods usually result in better predictive accuracy than filter methods.

Forward Feature Selection:
- This is an iterative method wherein we start with the best performing variable against the target. Next, we select another variable that gives the best performance in combination with the first selected variable. This process continues until the preset criterion is achieved.

Backward Feature Elimination:
- This method works exactly opposite to the Forward Feature Selection method. Here, we start with all the features available and build a model. Next, we the variable from the model which gives the best evaluation measure value. This process is continued until the preset criterion is achieved.

Exhaustive Feature Selection:
- This is the most robust feature selection method covered so far. This is a brute-force evaluation of each feature subset. This means that it tries every possible combination of the variables and returns the best performing subset.

Recursive Feature Elimination:
- ‘Given an external estimator that assigns weights to features (e.g., the coefficients of a linear model), the goal of recursive feature elimination (RFE) is to select features by recursively considering smaller and smaller sets of features. First, the estimator is trained on the initial set of features and the importance of each feature is obtained either through a coef_ attribute or through a feature_importances_ attribute.
- Then, the least important features are pruned from the current set of features. That procedure is recursively repeated on the pruned set until the desired number of features to select is eventually reached.’[2]

## Embedded methods

These methods encompass the benefits of both the wrapper and filter methods, by including interactions of features but also maintaining reasonable computational cost. Embedded methods are iterative in the sense that takes care of each iteration of the model training process and carefully extracts those features which contribute the most to the training for a particular iteration.

LASSO Regularization (L1):
- Regularization consists of adding a penalty to the different parameters of the machine learning model to reduce the freedom of the model, i.e. to avoid over-fitting. In linear model regularization, the penalty is applied over the coefficients that multiply each of the predictors. From the different types of regularization, Lasso or L1 has the property that is able to shrink some of the coefficients to zero. Therefore, that feature can be removed from the model.

Random Forest Importance:
- Random Forests is a kind of a Bagging Algorithm that aggregates a specified number of decision trees. The tree-based strategies used by random forests naturally rank by how well they improve the purity of the node, or in other words a decrease in the impurity (Gini impurity) over all trees. Nodes with the greatest decrease in impurity happen at the start of the trees, while notes with the least decrease in impurity occur at the end of trees. Thus, by pruning trees below a particular node, we can create a subset of the most important features.

# 19.	What is scaling? Where is it used? Which types do exist?

## What is scaling and where is it used

Feature Scaling is a technique to standardize the independent features present in the data in a fixed range. It is performed during the data pre-processing to handle highly varying magnitudes or values or units. If feature scaling is not done, then a machine learning algorithm tends to weigh greater values, higher and consider smaller values as the lower values, regardless of the unit of the values.

e.g. an algorithm with consider 3000 meters to be greater than 5 km, which results in wrong predictions. Feature scaling brings all values to the same magnitude, tackling the issue.

## Techniques for scaling

Normalization (sklearn: MinMaxScaler):
- This technique re-scales a feature or observation value with distribution value between 0 and 1.
- Only required when the data distribution is unknown or the data doesn't have Gaussian Distribution. This type of scaling technique is used when the data has a diversified scope and the algorithms on which the data are being trained do not make presumptions about the data distribution

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*mHmxR4CMM7aQFcrX1WvpNw.jpeg)

Standardization (sklearn: StandardScaler):
- It is a very effective technique which re-scales a feature value so that it has distribution with 0 mean value and variance equals to 1.
- Usually preferred when the data is being used for multivariate analysis i.e. when we want all the variables of comparable units. It is usually applied when the data has a bell curve i.e. it has gaussian distribution. No this isn't always true but is considered more effective when applied to Gaussian distribution. This technique comes in handy when the data has varying ratios and the algorithms used, make assumptions about the data distribution

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*-V6TGYz2E-kkzgxGuwmnpg.jpeg)

Robust Scalar:
- Robust scaling is one of the best scaling techniques when we have outliers present in our dataset. It scales the data accordingly to the interquartile range (IQR = 75 Quartile — 25 Quartile).

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*mwdkncyXehKKkmWBs9-ECg.png)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*HW7-kYjj6RKwrO-5WTLkDA.png)


# 20	How would you visualize multidimensional data

## 1 Dimension
- histograms
- density plot
- bar plots
- pie-charts

## 2 Dimensions
- pair-wise correlation matrix as heatmap
- pair-wise scatter plots
- parallel coordinates
- joint plots
- stacked bars or multiple bars
- box plots
- violin plots
- faceting\subplots along with generic histograms or density plots.


To accomodate more than 2 dimensions introduce:
- 3 axis (length, breadth, depth)
- Kernel density plot

Maxing out dimensions in seaborn:
- x="flipper_length_mm", 
- y="bill_length_mm",
- kind="scatter",
- hue='species',
- style= 'sex',
- size = "body_mass_g",
- col = 'island')

![](https://seaborn.pydata.org/_images/function_overview_8_0.png)

# 21.	What is over/under fitting, and how can you avoid It?


Underfitting:
- High training error
- Training error close to test error
- High bias
-- Bias Error: The simplifying assumptions made by a model to make the target function easier to learn. Generally, simple parametric algorithms have a high bias making them fast to learn and easier to understand but generally less flexible.

Overfitting:
- Very low training error
- Training error much lower than test error
- High variance
-- Variance Error: Variance is the amount that the estimate of the target function will change if different data was used. Ideally, it should not change too much from one training dataset to the next, meaning that the algorithm is good at picking out the hidden underlying mapping between the inputs and the output variables.



![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*9DtauyXaQFAOi31Pp1XQUg.png)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/72e222c1542539754df1d914cb671bd7.png)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Bias-and-Variance-Model.png)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-06%20at%2016.51.52.png)

# 22.	What are ‘training Set’ and ‘test Set’ in a Machine Learning Model? How Much Data Will You Allocate for Your Training, Validation, and Test Sets?


In most supervised machine learning tasks, best practice recommends to split your data into three independent sets:
- Training Set: For training of the model. 
- Validation Set: For unbiased evaluation of the model.
- Test Set: For final evaluation of the model. 

Training Set:
- The dataset that we feed our model to learn potential underlying patterns and relationships.
- The training set should be as representative as possible of the population that we are trying to model. Additionally, we need to be careful and ensure that it is as unbiased as possible, as any bias at this stage may be propagated downstream during inference.

Model building:
- Building different models to classify/predict based on different features
- We could compare the accuracy of each model on the training set, but if we use the same exact dataset for both training and tuning, the model will overfit and won't generalize well.

Validation Set:
- The dataset that we use to understand our model's performance across different model types and hyperparameter choices.
- it acts as an independent, unbiased dataset for comparing the performance of different algorithms trained on our training set. 
- Used to identify the best model

Test Set:
- The dataset that we use to approximate our model's unbiased accuracy in the wild.
- Once we have used the validation set to determine the algorithm and parameter choices that we would like to use in production, the test set is used to approximate the models's true performance in the wild. It is the final step in evaluating our model's performance on unseen data. 
- We should never, under any circumstance, look at the test set's performance before selecting a model. 

Peeking at our test set performance ahead of time is a form of overfitting, and will likely lead to unreliable performance expectations in production. It should only be checked as the final form of evaluation, after the validation set has been used to identify the best model.


https://mlu-explain.github.io/train-test-validation/

# Additional stuff

## Pickle

Allows you to save a trained model to share, commit, and re-load it faster.

1. Load data
2. Train model -  split into test/train, instantiate model and fit
3. Save the model pickle.dump(model, open('model.pkl', 'wb'))
4. Load the model pickled_model = pickle.load(open('model.pkl', 'rb'))
5. Use the pre-trained model to predict new data pickled_model.predict(X_test)

Alternative is to save as json files

Machine learning models often take hours or days to run, especially on large datasets with many features. If your machine goes off, you’ll lose your model and you’ll need to re-train it from scratch.



## Feature importance / model explainability

yellowbrick FeatureImportances

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2014.58.21.png)

ELI5

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2014.58.48.png)

LIME

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2015.01.00.png)

SHAP

shap.summary_plot

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2015.02.00.png)

shap.summary_plot (type=bar)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2015.42.35.png)

shap.dependence_plot

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2015.42.56.png)

shap.initsj

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Screenshot%202022-10-09%20at%2015.02.16.png)

Elbow method:

- x = number of clusters / features in PCA
- Cumulative variance explainability / percent of variance explaned
-- K-means score / sum of squared distances

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*8wV1j-klQA1xFvfaNXuVzg.png)

## SML in lecture notebooks

Regression problem:
1. Import data
2. Specify target (y)
3. Check out the correlation matrix and pairplot
4. Preprocess data - Standardscaling (instantiate, fit, transform)
5. Preprocess data - split into test and training set
6. Instantiate the models
7. Fit the models to training data (different models with increasing complexity and hyperparameter tuning options
- OLS model (baseline)
- Elastic Net (still parametric, but maybe advantages in feature selection)
- Random forest regressor (tree-based ensemble model)
8. Score the models on the test data using .score
9. Tune hyperparameters using grid search and scorer as the scoring method. Fit grid to X and y and return the best estimator




When using scaling models:
1. Instantiate the model
2. Fit the data onto the model (The fit function computes the formulation to transform the column based on Standard scaling but doesn’t apply the actual transformation. The computation is stored as a fit object. The fit method doesn’t return anything.)
3. Transform the data onto the dataset (The transform method takes advantage of the fit object in the fit() method and applies the actual transformation onto the column. So, fit() and transform() is a two-step process that completes the transformation in the second step. Here, Unlike the fit() method the transform method returns the actually transformed array.)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/metrics.png)

## Four types of data
Nominal
- categorical, equally good (cities, colours, gender)

Ordinal
- categorical, not equally good ([high, medium, low], [graduate, masters, phd])

Discrete
- countable integer values (number of students in a class, days in a week)

Continous
- Fractinoal float values (height, speed, temperature)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/types-of-data-.png)

## Overall ML process

1.	Loading data and pre-processing dataset of interest
2.	Hyperparameter optimization using cross-validation
3.	Fitting tuned algorithm to the training data
4.	Applying learned model to test data

## Cross-validation
Cross-validation is a resampling method of evaluating the validity of an ML model using a data sample. A technique that lets one to weigh the overfitting or underfitting extent of a model using the training data and testing data, cross-validation also allows one to test the accuracy of a model before launching it for public use. 
 
While there are numerous ways to cross-validate a model, all methods aim to test a data sample kept aside from the training data set.
1.	Separate a part of the data set from the rest.
2.	Utilize the rest of the data set to train the ML model.
3.	Once the model is ready, validate the model using the data set that was separated earlier.
 
Types of cross-validation:
1.	Holdout Cross-Validation
2.	K-fold Cross-Validation
3.	Stratified K-fold Cross-Validation
4.	Leave-p-out Cross-Validation
5.	Leave-one-out Cross-Validation
6.	Rolling Cross-Validation
7.	Monte Carlo Cross-Validation

## Grid Search / hyperparameter tuning / Regularization
Grid-search is used to find the optimal hyperparameters of a model which results in the most ‘accurate’ predictions. Grid search can be used to improve any specific evaluation metric.

The process of minimizing bias and variance by selecting
•	Model architecture
•	Learning rate
•	Number of epochs
•	Number of branches in a decision tree
•	Number of clusters in a clustering algorithm
•	The right model class
•	Functional form
•	Degree of complexity

## K-fold crossvalidation
Advanced version of Out-of-Sample-Testing, where the process is repeated several times during the hyperparameter-tuning phase

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/Picture%201.png)



## Distances
Cosine - used in collaborative recommendation systems

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/1*FTVRr_Wqz-3_k6Mk6G4kew.png)

![](https://raw.github.com/NadiaHolmlund/M1_exam/main/cosine-similarity-vectors.original.jpg)